infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/AsyncStoragePodInterceptor.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.openshift.provision;
+
+import static java.lang.String.format;
+import static org.eclipse.che.workspace.infrastructure.kubernetes.namespace.pvc.CommonPVCStrategy.COMMON_STRATEGY;
+import static org.eclipse.che.workspace.infrastructure.kubernetes.namespace.pvc.EphemeralWorkspaceUtility.isEphemeral;
+import static org.eclipse.che.workspace.infrastructure.openshift.provision.AsyncStorageProvisioner.ASYNC_STORAGE;
+
+import io.fabric8.kubernetes.api.model.DoneablePod;
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.Watch;
+import io.fabric8.kubernetes.client.Watcher;
+import io.fabric8.kubernetes.client.dsl.PodResource;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import javax.inject.Inject;
+import javax.inject.Named;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesInfrastructureException;
+import org.eclipse.che.workspace.infrastructure.openshift.OpenShiftClientFactory;
+import org.eclipse.che.workspace.infrastructure.openshift.environment.OpenShiftEnvironment;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This interceptor checks whether the starting workspace is configured with persistent storage and
+ * makes sure to stop the async storage pod (if any is running) to prevent "Multi-Attach error for
+ * volume". After the async storage pod is stopped and deleted, the workspace start is resumed.
+ */
+public class AsyncStoragePodInterceptor {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AsyncStoragePodInterceptor.class);
+  private static final int DELETE_POD_TIMEOUT_IN_MIN = 5;
+
+  private final OpenShiftClientFactory clientFactory;
+  private final String strategy;
+
+  @Inject
+  public AsyncStoragePodInterceptor(
+      @Named("che.infra.kubernetes.pvc.strategy") String strategy,
+      OpenShiftClientFactory openShiftClientFactory) {
+    this.strategy = strategy;
+    this.clientFactory = openShiftClientFactory;
+  }
+
+  public void intercept(OpenShiftEnvironment osEnv, RuntimeIdentity identity)
+      throws InfrastructureException {
+    if (!COMMON_STRATEGY.equals(strategy) || isEphemeral(osEnv.getAttributes())) {
```

Comment:
@vparfonov wondering if this is not checked already on the `AsyncStorageModeValidator.java ` level?